packages/gatsby-plugin-sitemap/src/gatsby-node.js

```diff
@@ -1,80 +1,148 @@
 import path from "path"
-import sitemap from "sitemap"
+import minimatch from "minimatch"
+import { simpleSitemapAndIndex } from "sitemap"
+import { validateOptions } from "./options-validation"
 import {
-  defaultOptions,
-  filterQuery,
-  writeFile,
-  renameFile,
   withoutTrailingSlash,
+  prefixPath,
+  defaultFilterPages,
+  defaultExcludes,
 } from "./internals"
 
-const publicPath = `./public`
+const PublicPath = `./public`
+
+const ReporterPrefix = `[gatsby-plugin-sitemap]:`
+
+exports.onPreInit = async ({ reporter }, pluginOptions) => {
+  try {
+    await validateOptions(pluginOptions)
+    reporter.verbose(`${ReporterPrefix} Plugin options passed validation.`)
+  } catch (err) {
+    reporter.panic(err)
+  }
+}
 
 exports.onPostBuild = async (
-  { graphql, pathPrefix, basePath = pathPrefix },
+  { graphql, reporter, pathPrefix },
   pluginOptions
 ) => {
-  const options = { ...pluginOptions }
-  delete options.plugins
-  delete options.createLinkInHead
-
+  // Schema was already validated in preInit but we use joi to get our pluginOptions with default options.
   const {
-    query,
-    serialize,
     output,
-    exclude,
-    hostname,
+    entryLimit,
+    query,
+    excludes,
     resolveSiteUrl,
-    ...rest
-  } = {
-    ...defaultOptions,
-    ...options,
-  }
+    resolvePagePath,
+    resolvePages,
+    filterPages,
+    serialize,
+  } = await validateOptions(pluginOptions).catch(err => {
+    reporter.panic(err)
+  })
 
-  const saved = path.join(publicPath, output)
+  const { data: queryRecords } = await graphql(query)
 
-  // Paths we're excluding...
-  const excludeOptions = exclude.concat(defaultOptions.exclude)
+  reporter.verbose(
+    `${ReporterPrefix} Query Results:\n${JSON.stringify(queryRecords, null, 2)}`
+  )
 
-  const queryRecords = await graphql(query)
+  // resolvePages and resolveSuteUrl are allowed to be sync or async. The IIFE handles each possibility
+  const allPages = await (async () => resolvePages(queryRecords))().catch(err =>
+    reporter.panic(`Error resolving Pages`, err)
+  )
 
-  const filteredRecords = filterQuery(
-    queryRecords,
-    excludeOptions,
-    basePath,
-    resolveSiteUrl
+  const siteUrl = await (async () =>
+    resolveSiteUrl(queryRecords))().catch(err =>
+    reporter.panic(`Error resolving Site URL`, err)
```

Comment:
This is here to handle both sync and async functions while also catching errors. If feels messy but I can't think of a better way. I didn't like try/catch cause I didn't want to have to handle scope. Any suggestions or is this the way?